### ISSUE TITLE
Enable inference rollouts that require only initial condition + time-dependent forcing

### ISSUE THREAD

**Issue by oliverwm1** [ai2cm/ace#207]
It would be useful to be able to do arbitrarily long inference runs. Currently we are limited by the length of the validation dataset because the code assumes that complete target data will always be available for computing metrics. We should relax this to only require that the variables used for forcing (inputs only) be in the validation dataset, and simply not compute metrics when target data is not available.

**Comment by mcgibbon** [comment]
I think we also will need to modify the inference so it annually repeats the forcing data, right? We only have a finite amount of validation data.

**Comment by oliverwm1** [comment]
> I think we also will need to modify the inference so it annually repeats the forcing data, right? We only have a finite amount of validation data.

It would be straightforward to manually extend the forcing part of the validation data offline. I would be hesitant about implementing the annual repetition of forcing data within the inference code since annually-repeating is a special case that we won't always be using.

**Comment by mcgibbon** [comment]
Ah for a second I thought this is easy because we only need to update the metric calculations to work if there's no target, but it's more complicated because it affects the data loading code. We'll need to be able to load from datasets where variables have different time axis lengths (we can define them under different dimension names to allow this).

**Comment by oliverwm1** [comment]
> Ah for a second I thought this is easy because we only need to update the metric calculations to work if there's no target, but it's more complicated because it affects the data loading code. We'll need to be able to load from datasets where variables have different time axis lengths (we can define them under different dimension names to allow this).

I think we can impose some limitations to make it easier to implement. E.g. we could generate a validation dataset that only has the forcing data so that all variables have the same time axis length (although it will still need all prognostic variables for initial condition).

Or another approach: insist each "window" has variables which all have the same length.

**Comment by mcgibbon** [comment]
Because we need the IC we can't have everything with the same time axis length.

I'm sure this can be solved, it's just not as trivial as updating the metrics.

**Comment by gideonite** [comment]
It sounds like we will create forcing and initial conditions offline and store them on disk as inputs to the model. This will make actually running the model pretty straight forward. The challenge will come with the metrics.

The first question, I think, is to ask what we expect to log for such runs? The obvious answer would be gen-only metrics, no error or biases. In this case, do we care if there are empty fields in the wandb page for the missing metrics?

How do we feel about having yet another flag `gen_only: bool` or something like that which would have to match the ic & forcing-only inference dataset?

**Comment by oliverwm1** [comment]
> The first question, I think, is to ask what we expect to log for such runs? The obvious answer would be gen-only metrics, no error or biases. In this case, do we care if there are empty fields in the wandb page for the missing metrics?

Agreed w.r.t. what we would log. Don't think we need to worry about missing metrics—wandb seems quite resilient to comparing runs with different metrics available.

> How do we feel about having yet another flag gen_only: bool or something like that which would have to match the ic & forcing-only inference dataset?

I'm always in favor of fewer flags if possible :) Based on what is in the provided validation dataset, it seems the code should be able to determine which metrics it can compute.

Another approach would be to make explicit the distinction between initial conditions, forcing data and validation data. Right now these are all mixed up in what we call the "target data". E.g. could have users pass separate paths for initial condition data, forcing data and (optionally) validation data. But it would be more work for users to set up initial condition ensembles compared to our current setup so not sure we want to do this.

**Comment by mcgibbon** [comment]
I think the separate paths idea is neat.

**Comment by gideonite** [comment]
> I'm always in favor of fewer flags if possible :)

Yes me too :-)

> But it would be more work for users to set up initial condition ensembles compared to our current setup so not sure we want to do this.

I'm not sure I understand what the downside is for ic ensembles. Is it that practitioners would need to specify two files rather one?

**Comment by oliverwm1** [comment]
> I'm not sure I understand what the downside is for ic ensembles. Is it that practitioners would need to specify two files rather one?

I can imagine instances where doing an initial condition ensemble would require a new step of generating some set of initial condition files, whereas with the current code the ensemble could be done only by changing inference config.

Having separate files for initial condition/forcing/validation will also open the door to bugs related to the time coordinate not aligning properly between these three dataset. Not a blocker but will take some thought and care to get right. This is a major feature IMO—would be useful but also a fair bit of work to implement.

**Comment by gideonite** [comment]
The current logic around `n_samples`, `n_candidate_ics`, `n_initial_conditions`, etc. seems to be some indication that this is indeed a tricky issue. When `inference.n_forward_steps` is larger than the dataset size, the existing logic results in a DataLoader of length = 0. See below and ai2cm/full-model#360 for the failing test.

Error:

https://github.com/ai2cm/full-model/blob/6834b38c52089c70efaad30040c74d9e4a1b0611/fme/fme/fcn_training/inference/loop.py#L137-L138

Code: 

https://github.com/ai2cm/full-model/blob/6834b38c52089c70efaad30040c74d9e4a1b0611/fme/fme/core/data_loading/xarray.py#L126-L148

**Comment by gideonite** [comment]
Also, shouldn't the code be expected to run when there are no forcing variables at all?

**Comment by gideonite** [comment]
Jeremy and I just had a super helpful sync. Two PRs are probably going to emerge:

1. Modify `run_on_batch` to handle the case where target data doesn't exist.
2. Modify XarrayDataset so that there is an `initial_condition` axis in addition to a `time` axis. This will handle the problem of some variable (e.g. ic prognostic variables) only being present at `t = 0`.


Note that (1) puts forward a few challenges: (a) packing / unpacking data correctly, (b) how to modify the loss computation for the case when target_data doesn't exist (we've put some effort into avoiding the `train: bool` flag).

**Comment by mcgibbon** [comment]
Just to add a few explicit details from our conversation for clarity of anyone reading this, in (1) non-prescribed prognostic values provided to `run_on_batch` will have a length-1 time dimension. In (2), non-prescribed prognostic values may have an `initial_condition` axis (which we might at first assume to be length-1) in place of a "time" axis. If this happens, the data loader will produce windows only using this/these initial condition(s), with the forcing data kept identical for each initial condition (starting from the start of the time axis, which is taken as the initial condition for those variables).

**Comment by gideonite** [comment]
This is split into three PRs. The first is underway, the second is to do.

- [x] ai2cm/full-model#364 
- [x] ai2cm/full-model#376 
- [x] ai2cm/full-model#383 
- [x] Work out bugs from putting all of these commits together.

**Comment by gideonite** [comment]
> non-prescribed prognostic values may have an initial_condition axis (which we might at first assume to be length-1) in place of a "time" axis

Thinking about this a bit more, I'm not 100% convinced that this is the way to go. What if want to simply run inference beyond the 1-year validation dataset that we have on disk? In this formulation, the 1-year dataset would constitute 1-year of initial conditions which doesn't make sense to me. Maybe it would make more sense to add a forcing-related dimension? @mcgibbon

**Comment by mcgibbon** [comment]
This rollout will use a new dataset on disk. If we make this new dataset from the 1-year inference, we would do this by taking only the first time sample for variables having an `initial_condition` axis, and all time samples for variables where we want a `time` axis. We would not want to use the full dataset for initial conditions, because only the first time sample would be compatible with the "forcing" data (which starts at the first time sample).

The only time we'd do more than 1 initial condition is if we did some kind of perturbation study, where we slightly change the initial values, or some kind of ensemble equilibration study, where we expect a variety of initial conditions to converge to similar long-term climates and don't really look at the "spin-up" period.

**Comment by oliverwm1** [comment]
I don't think this is completed. We made some steps towards it, but AFAIK it is not possible to run without the full target data.

**Comment by gideonite** [comment]
I think that this technically can be achieved by repeating the initial conditions for the number of desired steps.
